core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1163,6 +1176,11 @@ public void failShard(String reason, @Nullable Throwable e) {
         public ShardRouting routingEntry() {
             return indexShard.routingEntry();
         }
+
+        @Override
+        public long opPrimaryTerm() {
+            return routingEntry().primaryTerm();
+        }
```

Comment:
I wonder what our system invariants are w.r.t primaryTerm in IndexShard. Primary term has for example a direct impact on the reroute phase. A new primary term means that we need to route to a new primary location. The big question is: What is the relation between primaryTerm that we use for routing to primaryTerm in IndexShard? Maybe we could require both to be the same and otherwise abort primaryOperation?

Other invariants to think about:
- If IndexShard gets created as primary shard (that is not a relocation target), then it should have the same primary term until its closed, right?
- If IndexShard gets created as primary shard (that IS a relocation target), then it should switch its primary term only once, namely when it gets activated.

I wonder how can we better capture these invariants in the code.